streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java

```diff
@@ -123,7 +123,7 @@ public void shouldAllowJoinMaterializedFilteredKTable() {
 
         assertThat(
             topology.stateStores().size(),
-            equalTo(1));
+            equalTo(2));
```

Comment:
This is surprising to me -- need to dig into the code I guess, but I though enabling to send old values would _not_ trigger any upstream materialization but only take effect _if_ the store is materializes already (but would be ignored otherwise). Seems my assumption is wrong though.

Reply:
The issue seems to be this line: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableSource.java#L67

In a source-KTable, if we enable sending old values we force an materialization. I guess, we should have this materialization optional?

I would recommend to rename `enableSendingOldValues()` to `maybeEnableSendingOldValues(final boolean forceMaterialization)` or similar?

\cc @guozhangwang WDYT?

Reply:
`enableSendingOldValues` is indicated for sending a pair of <new, old> values, and if it is not set the `old` value would always be null. If we want to `enableSendingOldValues` then we'd have to materialize the source node still?

Maybe I'm not fully understanding the context here. Could you bring me up to date?

Reply:
@guozhangwang Atm, when `enableSendingOldValues` is set and the upstream store is not materialized already we enforce a materialization. Thus, enabling `enableSendingOldValues` in `KTable#filter()` would be a breaking change as we would start to materialize state that did not exist before if one upgrades a topology.

Instead, we want to say, _iff_ the upstream store exist, please send me the old value, but if the upstream store does not exist, it's ok to just send `old=null` but don't force a materialization.

Reply:
Ah I see.. what about changing the function as `enableSendingOldValues(boolean enforced)` in that case? And then only ktable aggregations which would definitely need this would set it to `true`.

Reply:
Sounds good.